### ISSUE TITLE
add 1 or more complexity metrics for the power law fits

### ISSUE THREAD

**Issue by charlesmartin14** [CalculatedContent/WeightWatcher#173]
What is the best way to evaluate the absolute quality and/or complexity of the 1 dimensional function.  That is, is there a metric that tells us if the data is actually power law, not by simply measuring the Kolmogorov–Smirnov distance , but with some other metric that takes into account the non-convex nature of the 1 dimensional function ?

-- from ChatGPT

The entropy of a 1-dimensional function can be measured using the following steps:

Discretize the function: The first step is to discretize the function by dividing the range of the function into a set of bins. This is done to convert the continuous function into a discrete function.

Compute the probability density of the function: The next step is to compute the probability density of the function in each bin. The probability density of a bin is the number of data points in the bin divided by the total number of data points.

Compute the entropy of the function: Once the probability density of the function is computed, the entropy of the function can be computed as the sum of the probability density of each bin multiplied by the logarithm of the probability density. The entropy is computed as:

S = - ∑ p(x) * log(p(x))

Where p(x) is the probability density of the function in bin x.

The entropy of the function is a measure of the disorder in the function. The entropy will be low if the probability density of the function is concentrated in a few bins, and high if the probability density of the function is spread out across many bins.
It's worth mentioning that the entropy computation can be done in multiple ways like using the Kullback-Leibler divergence or Renyi entropy. Additionally, the discretization of the function can also be done in multiple ways such as using a fixed bin width or use a variable bin width depending on the data set.

Is there a python function in say scikit learn or scipy that can do this for an arbitrary 1 d function
Scikit-learn and Scipy libraries do not have a built-in function that can compute the entropy of an arbitrary 1-dimensional function directly. However, you can use the functions in these libraries to implement the steps required to compute the entropy of a 1-dimensional function.

In scipy, you can use the histogram function to discretize the function, and the entropy function from the stats module to compute the entropy of the function.

In scikit-learn, you can use the KBinsDiscretizer class to discretize the function, and then use the entropy_estimator function from the metrics module to compute the entropy of the function.

Additionally, you can use the numpy library to implement the steps of discretization and entropy calculation.

It's important to note that the entropy computation will be influenced by the discretization method and parameters used, thus it's a good idea to test the algorithm with different bin widths and compare the results.

Act like a python programmer .  Create for me a function that can take as input a 1 dimensional array, and measure its entropy, using the method you just described.  
Sure, here's a Python function that takes a 1-dimensional array as input and computes its entropy using the method I described earlier:

Copy code
import numpy as np

def compute_entropy(data):
    # Discretize the data
    hist, bin_edges = np.histogram(data, bins='auto')
    # Compute the probability density of the function
    prob = hist / np.sum(hist)
    # Compute the entropy of the function
    entropy = -np.sum(prob * np.log2(prob))
    return entropy
This function takes a 1-dimensional array as input, and it uses the numpy library to discretize the data and compute the entropy. The data is discretized by using the 'auto' option in the numpy histogram function, which automatically chooses the bin width. You can also set a specific number of bins, or pass an array of bin edges to use for the discretization.

You can use this function by calling it and passing the 1-dimensional array as an argument, like this:

Copy code
my_data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
entropy = compute_entropy(my_data)
print(entropy)
Also, you can use the sklearn.metrics.entropy function to calculate the entropy of the function using the probability density of the function.

Copy code
from sklearn.metrics import entropy

def compute_entropy(data):
    # Discretize the data
    hist, bin_edges = np.histogram(data, bins='auto')
    # Compute the probability density of the function
    prob = hist / np.sum(hist)
    # Compute the entropy of the function
    entropy = entropy(prob)
    return entropy
Please note that the function above will work with the assumption that the data passed is 1D and not empty, otherwise, you might want to add some control flow

`from sklearn.metrics import entropy

def compute_entropy(data):
    # Discretize the data
    hist, bin_edges = np.histogram(data, bins='auto')
    # Compute the probability density of the function
    prob = hist / np.sum(hist)
    # Compute the entropy of the function
    entropy = entropy(prob)
    return entropy`

**Comment by charlesmartin14** [comment]
OMFG I think ChatGPT made up the entropy function in sklearn 📄

**Comment by charlesmartin14** [comment]
correct code

```
from scipy.stats import entropy

def compute_entropy(data):
    # Discretize the data
    hist, bin_edges = np.histogram(data, bins='auto')
    # Compute the probability density of the function
    prob = hist / np.sum(hist)
    # Compute the entropy of the function
    the_entropy = entropy(prob)
    return the_entropy
```

![image](https://user-images.githubusercontent.com/498448/214753422-a20706b1-1b15-4337-9aff-10f1a40c843e.png)

**Comment by charlesmartin14** [comment]
Entropy does appear to be a different metric than alpha, D, and sigma:

![image](https://user-images.githubusercontent.com/498448/214754575-436624fb-3c15-460b-889c-40a901bd5aec.png)

![image](https://user-images.githubusercontent.com/498448/214754587-4eb03b87-6c99-40d9-9710-effd00e6c573.png)

![image](https://user-images.githubusercontent.com/498448/214754594-e385aa83-3076-41a6-b9d0-3c1470cad28f.png)

**Comment by charlesmartin14** [comment]
Its been suggested that this measure of the entropy does not capture the flat nature of the plot in layer 17, above
open tp other ideas
